src/System.Text.Primitives/System/Text/Formatting/Internal/FormattingHelpers.cs

```diff
@@ -0,0 +1,144 @@
+        public static int WriteFractionDigits(long value, int digitCount, ref byte buffer, int index)
+        {
+            for (var i = FractionDigits; i > digitCount; i--)
+                value = DivMod(value, 10, out long m);
+
+            return WriteDigits(value, digitCount, ref buffer, index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteDigits(long value, ref byte buffer, int index)
+        {
+            return WriteDigits(value, CountDigits(value), ref buffer, index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteDigits(long value, int digitCount, ref byte buffer, int index)
+        {
+            long left = value;
+
+            for (var i = digitCount - 1; i >= 0; i--)
+            {
+                left = DivMod(left, 10, out long num);
+                Unsafe.Add(ref buffer, index + i) = (byte)('0' + num);
+            }
+
+            return digitCount;
+        }
+
+        #endregion UTF-8 Helper methods
+
+        #region UTF-16 Helper methods
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void WriteHexByte(byte value, ref char buffer, int index)
+        {
+            Unsafe.Add(ref buffer, index) = HexTable[value >> 4];
+            Unsafe.Add(ref buffer, index + 1) = HexTable[value & 0xF];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteFractionDigits(long value, int digitCount, ref char buffer, int index)
+        {
+            for (var i = FractionDigits; i > digitCount; i--)
+                value = DivMod(value, 10, out long m);
+
+            return WriteDigits(value, digitCount, ref buffer, index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteDigits(long value, ref char buffer, int index)
+        {
+            return WriteDigits(value, CountDigits(value), ref buffer, index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteDigits(long value, int digitCount, ref char buffer, int index)
+        {
+            long left = value;
+
+            for (var i = digitCount - 1; i >= 0; i--)
+            {
+                left = DivMod(left, 10, out long num);
+                Unsafe.Add(ref buffer, index + i) = (char)('0' + num);
+            }
+
+            return digitCount;
+        }
+
+        #endregion UTF-16 Helper methods
+
+        #region Math Helper methods
+
+        /// <summary>
+        /// We don't have access to Math.DivRem, so this is a copy of the implementation.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static long DivMod(long numerator, long denominator, out long modulo)
+        {
+            long div = numerator / denominator;
+            modulo = numerator - (div * denominator);
```

Comment:
The % operator is *really* expensive. The compiler should do this for us (like the C++ compiler does), but it doesn't. This trick is actually coded into the Math library as a DivRem method, but I don't have access to it here. My testing shows this to be significantly faster.